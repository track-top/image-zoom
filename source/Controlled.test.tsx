import { act, fireEvent, render } from '@testing-library/react'
import React from 'react'
import { renderToString } from 'react-dom/server'
import ControlledZoom from './Controlled'

test(`when closed and then open`, () => {
  jest.useFakeTimers()

  const handleZoomChange = jest.fn()
  const { getByLabelText, getByRole, rerender } = render(
    <ControlledZoom isZoomed={false} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()

  fireEvent.click(openTrigger)
  expect(handleZoomChange).toHaveBeenLastCalledWith(true)

  rerender(
    <ControlledZoom isZoomed={true} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  expect(document.body).toMatchSnapshot()

  // nothing should happen
  fireEvent.click(openTrigger)

  const closeTrigger = getByLabelText(`Unzoom image`)
  const modal = getByRole(`dialog`)
  expect(closeTrigger).toBeVisible()

  fireEvent.click(closeTrigger)
  expect(handleZoomChange).toHaveBeenLastCalledWith(false)

  rerender(
    <ControlledZoom isZoomed={false} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  expect(document.body).toMatchSnapshot()

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(closeTrigger).not.toBeInTheDocument()
  expect(modal).not.toBeInTheDocument()
  expect(document.body).toMatchSnapshot()
})

test(`when open and then closed`, () => {
  const handleZoomChange = jest.fn()
  const { getByLabelText, getByRole, rerender } = render(
    <ControlledZoom isZoomed={true} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  const closeTrigger = getByLabelText(`Unzoom image`)
  const modal = getByRole(`dialog`)
  expect(closeTrigger).toBeVisible()

  fireEvent.click(closeTrigger)
  expect(handleZoomChange).toHaveBeenLastCalledWith(false)

  rerender(
    <ControlledZoom isZoomed={false} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  expect(document.body).toMatchSnapshot()

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(closeTrigger).not.toBeInTheDocument()
  expect(modal).not.toBeInTheDocument()

  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()
})

test(`sends unzoom message when ESC key pressed`, () => {
  const handleZoomChange = jest.fn()
  const { getByRole } = render(
    <ControlledZoom isZoomed={true} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  const modal = getByRole(`dialog`)
  expect(modal).toBeVisible()

  // should do nothing
  fireEvent.keyDown(document, { key: `ArrowLeft` })
  expect(handleZoomChange).not.toHaveBeenCalled()

  // just in case
  expect(modal).toBeVisible()

  fireEvent.keyDown(document, { key: `Escape` })
  expect(handleZoomChange).toHaveBeenLastCalledWith(false)
})

test(`sends unzoom message when scrolled`, () => {
  const handleZoomChange = jest.fn()
  const { getByRole, rerender } = render(
    <ControlledZoom isZoomed={true} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  const modal = getByRole(`dialog`)
  expect(modal).toBeVisible()

  expect(handleZoomChange).not.toHaveBeenCalled()

  act(() => {
    window.dispatchEvent(new Event(`scroll`, {}))
  })

  rerender(
    <ControlledZoom isZoomed={false} onZoomChange={handleZoomChange}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  act(() => {
    // run scroll again to emulate actual scroll events firing
    window.dispatchEvent(new Event(`scroll`, {}))
    jest.advanceTimersByTime(300)
  })

  expect(handleZoomChange).toHaveBeenLastCalledWith(false)
})

test(`custom open/close text`, () => {
  const { getByLabelText, rerender } = render(
    <ControlledZoom
      closeText="Close me"
      isZoomed={false}
      onZoomChange={jest.fn()}
      openText="Open me"
    >
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  const openTrigger = getByLabelText(`Open me`)
  expect(openTrigger).toBeVisible()
  expect(document.body).toMatchSnapshot()

  rerender(
    <ControlledZoom
      closeText="Close me"
      isZoomed={true}
      onZoomChange={jest.fn()}
      openText="Open me"
    >
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  const closeTrigger = getByLabelText(`Close me`)
  expect(closeTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()
})

test(`renders without browser environment`, () => {
  const html = renderToString(
    <ControlledZoom isZoomed={false} onZoomChange={jest.fn()}>
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  document.body.innerHTML = html
  expect(document.body).toMatchSnapshot()
})

test(`renders with different props`, () => {
  const html = renderToString(
    <ControlledZoom
      isZoomed={false}
      overlayBgColorEnd="rgba(255, 255, 255, 0.95)"
      overlayBgColorStart="rgba(255, 255, 255, 0)"
      transitionDuration={400}
      wrapElement="span"
      zoomMargin={1}
      zoomZindex={2147483648}
    >
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledZoom>
  )

  document.body.innerHTML = html
  expect(document.body).toMatchSnapshot()
})
