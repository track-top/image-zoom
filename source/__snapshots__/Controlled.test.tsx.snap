// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom open/close text 1`] = `
<body>
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Open me"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`custom open/close text 2`] = `
<body>
  <div>
    <div
      aria-modal="true"
      data-rmiz-overlay="true"
      role="dialog"
      style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
    >
      <div
        data-rmiz-modal-content="true"
        style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
      </div>
      <button
        aria-label="Close me"
        data-rmiz-btn-close="true"
        type="button"
      />
    </div>
  </div>
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Open me"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`renders with different props 1`] = `
<body>
  <span
    data-rmiz-wrap="visible"
  >
    <img
      alt="foo"
      src="foo.jpg"
      width="500"
    />
    <button
      aria-label="Zoom image"
      data-rmiz-btn-open="true"
      type="button"
    />
  </span>
</body>
`;

exports[`renders without browser environment 1`] = `
<body>
  <div
    data-rmiz-wrap="visible"
  >
    <img
      alt="foo"
      src="foo.jpg"
      width="500"
    />
    <button
      aria-label="Zoom image"
      data-rmiz-btn-open="true"
      type="button"
    />
  </div>
</body>
`;

exports[`when closed and then open 1`] = `
<body>
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`when closed and then open 2`] = `
<body>
  <div>
    <div
      aria-modal="true"
      data-rmiz-overlay="true"
      role="dialog"
      style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
    >
      <div
        data-rmiz-modal-content="true"
        style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
      </div>
      <button
        aria-label="Unzoom image"
        data-rmiz-btn-close="true"
        type="button"
      />
    </div>
  </div>
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`when closed and then open 3`] = `
<body>
  <div>
    <div
      aria-modal="true"
      data-rmiz-overlay="true"
      role="dialog"
      style="background-color: rgba(255, 255, 255, 0); transition-duration: 300ms; z-index: 2147483647;"
    >
      <div
        data-rmiz-modal-content="true"
        style="height: 0px; left: 0px; top: 0px; transform: scale(1) translate(0, 0); transition-duration: 300ms; width: 0px;"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
      </div>
      <button
        aria-label="Unzoom image"
        data-rmiz-btn-close="true"
        type="button"
      />
    </div>
  </div>
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`when closed and then open 4`] = `
<body>
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`when open and then closed 1`] = `
<body>
  <div>
    <div
      aria-modal="true"
      data-rmiz-overlay="true"
      role="dialog"
      style="background-color: rgba(255, 255, 255, 0); transition-duration: 300ms; z-index: 2147483647;"
    >
      <div
        data-rmiz-modal-content="true"
        style="height: 0px; left: 0px; top: 0px; transform: scale(1) translate(0, 0); transition-duration: 300ms; width: 0px;"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
      </div>
      <button
        aria-label="Unzoom image"
        data-rmiz-btn-close="true"
        type="button"
      />
    </div>
  </div>
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`when open and then closed 2`] = `
<body>
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
        type="button"
      />
    </div>
  </div>
</body>
`;
